### PR TITLE
begin work on some homotopy coherences

### PR DESCRIPTION
**Pull request by jonalfcam**
In the course of proving 9.10 I ran into a small part of the proof that needs more homotopy coherences than I originally thought. I've begun putting in the ones that I need, but I figure it would be a good thing to leave most of the groupoidal structure for homotopies to someone as a first issue. What doe you think @emilyriehl @fizruk ? 

I also added a dependent homotopy type, which as far as I could tell was not in there?  Of course one can get away with $\lambda$-abstraction, but I found using this easier to read when working with homotopies between homotopies.

### CODE REVIEW

**Comment by jonalfcam** [comment]
As a side issue about documentation, the little lemma I need to prove is absolutely unparseable with a commutative diagram or some diagram accompanying it . Is there a way to include those?

**Comment by fizruk** [comment]
> As a side issue about documentation, the little lemma I need to prove is absolutely unparseable with a commutative diagram or some diagram accompanying it . Is there a way to include those?

There should be a way to enable [`amscd` package](https://www.jmilne.org/not/Mamscd.pdf) for MathJax, but I haven't yet figured out the settings. I will let you know when I have a working example.

Otherwise, we could attach pictures :)

File: src/hott/02-homotopies.rzk.md [hunk 1]
```diff
@@ -41,10 +41,58 @@ This is a literate `rzk` file:
...
 
+There is also a dependent version of homotopy.
+
+```rzk
+#def dependent-homotopy
+  ( A : U)
+  ( B : A → U)
+  (f g : (a : A) → B a)
+  : U
+  := (a : A) → (f a = g a)
```
Comment (fredrik-bakke) [inline src/hott/02-homotopies.rzk.md:65]: What you call a dependent homotopy here I would call just a homotopy. The notion of dependent homotopy should generalize dependent identifications. You can find a formalization of it here:

https://github.com/UniMath/agda-unimath/blob/591dd732c2bf6876871b941e952f88364c43ffe1/src/foundation/dependent-homotopies.lagda.md

Comment (jonalfcam) [reply src/hott/02-homotopies.rzk.md:65]: Indeed! But in rzk currently homotopies are not defined for dependent functions.

**Comment by fizruk** [comment]
> > > As a side issue about documentation, the little lemma I need to prove is absolutely unparseable with a commutative diagram or some diagram accompanying it . Is there a way to include those?
> > 
> > 
> > There should be a way to enable [`amscd` package](https://www.jmilne.org/not/Mamscd.pdf) for MathJax, but I haven't yet figured out the settings. I will let you know when I have a working example.
> > Otherwise, we could attach pictures :)
> 
> I’m happy to generate svg commutative diagrams! Just wanted to make sure it would be ok.

Yes, I believe it is okay to generate and add SVG commutative diagrams :)

File: src/hott/02-homotopies.rzk.md (hunk 1 above)
Comment (jonalfcam) [reply src/hott/02-homotopies.rzk.md:65]: @fredrik-bakke  Should I be the one to apply these changes globally? I'm happy to, but wanted to make sure.

**Comment by jonalfcam** [comment]
Added a commutative diagram.  It would maybe be nice to have a way to collapse these things in vscode? I have no idea how that works.

**Comment by fizruk** [comment]
> Added a commutative diagram. It would maybe be nice to have a way to collapse these things in vscode? I have no idea how that works.

If you indent the contents of `<svg>`, then you would be able to fold it by clicking the button to the left of `<svg>` tag:
Screenshot 2023-10-13 at 02 05 09


Screenshot 2023-10-13 at 02 05 30

File: src/hott/02-homotopies.rzk.md (hunk 1 above)
Comment (fredrik-bakke) [reply src/hott/02-homotopies.rzk.md:65]: That would be great, if you're comfortable with that

**Comment by jonalfcam** [comment]
Haha. Fantastic! I'm sorry for not knowing such an easy fix!

File: src/hott/02-homotopies.rzk.md (hunk 1 above)
Comment (jonalfcam) [reply src/hott/02-homotopies.rzk.md:65]: Yup! Absolutely.

File: src/hott/02-homotopies.rzk.md [hunk 2]
```diff
@@ -41,10 +41,58 @@ This is a literate `rzk` file:
...
+```
+
+## Some path algebra for homotopies
+
+This section contains some lemmas on the groupoidal structure of homotopies.
```
Comment (fredrik-bakke) [inline src/hott/02-homotopies.rzk.md:70]:
```suggestion
In this section we prove some lemmas on the groupoidal structure of homotopies.
```

File: src/hott/02-homotopies.rzk.md [hunk 3]
```diff
@@ -41,10 +41,58 @@ This is a literate `rzk` file:
...
+    (\ x → concat B (g x) (f x) (g x) (rev B (f x) (g x) (K x)) (H x))
+    (refl-htpy A B g)
+  := \ x →
+      cancel-left-path
+      (B)
+      (f x)
+      (g x)
+      (H x)
+      (K x)
+      (C x)
```
Comment (fredrik-bakke) [inline src/hott/02-homotopies.rzk.md:93]:
```suggestion
  := \ x → cancel-left-path B (f x) (g x) (H x) (K x) (C x)
```
When the code fits within the 80-character limit, it is usually preferable to write it on a single line.

File: src/hott/01-paths.rzk.md [hunk 4]
```diff
@@ -952,3 +965,151 @@ The following is the same as above but with alternating arguments.
...
+       ( associative-concat A v w v z p q r)
+       ( concat-eq-right
+        ( A )
+        ( v )
+        ( w)
+        ( z)
+        ( p)
+        ( concat A w v z q r)
+        ( concat A w y z s t)
+        ( H)))))
```
Comment (fredrik-bakke) [inline src/hott/01-paths.rzk.md:1114]:
```suggestion
    ( rev
      ( v = z)
      ( concat  A v v z refl r)
      ( r)
      ( left-unit-concat A v z r))
    ( concat
     ( v = z)
     ( concat A v v z refl r)
     ( concat A v v z (concat A v w v p q) r)
     ( concat A v w z p (concat A w y z s t))
     ( rev
       ( v = z)
       ( concat A v v z  (concat A v w v p q) r)
       ( concat A v v z refl r)
       ( concat-eq-left
         ( A)
         ( v)
         ( v)
         ( z)
         ( concat A v w v p q)
         ( refl)
         ( H')
         ( r)))
     ( concat
       ( v = z)
       ( concat A v v z (concat A v w v p q) r)
       ( concat A v w z p (concat A w v z q r))
       ( concat A v w z p (concat A w y z s t))
       ( associative-concat A v w v z p q r)
       ( concat-eq-right
         ( A)
         ( v)
         ( w)
         ( z)
         ( p)
         ( concat A w v z q r)
         ( concat A w y z s t)
         ( H)))))
```

If you're using VSCode, you can try the [indent-rainbow](https://marketplace.visualstudio.com/items?itemName=oderwat.indent-rainbow) plugin to help you spot misindented code.

File: src/hott/01-paths.rzk.md [hunk 5]
```diff
@@ -380,6 +380,19 @@ The following needs to be outside the previous section because of the usage of
...
+A special case of this is sometimes useful
+
+```rzk
+#def cancel-left-path
+  (A : U)
+  (x y : A)
+  (p q : x = y)
+  : (p = q) → ( concat A y x y (rev A x y q) p) = refl
+  := \ α →
+      triangle-rotation A x y y p q refl α
```
Comment (fredrik-bakke) [inline src/hott/01-paths.rzk.md:392]:
```suggestion
  := \ α → triangle-rotation A x y y p q refl α
```

File: src/hott/02-homotopies.rzk.md [hunk 6]
```diff
@@ -41,10 +41,58 @@ This is a literate `rzk` file:
...
+#def htpy-cancel-left
+ (A B : U)
+ (f g : A → B)
+ (H K : homotopy A B f g)
+ (C : dependent-homotopy A (\ a → f a = g a) H K)
+ : dependent-homotopy
+    A
+    (\ b → g b = g b)
+    (\ x → concat B (g x) (f x) (g x) (rev B (f x) (g x) (K x)) (H x))
+    (refl-htpy A B g)
```
Comment (fredrik-bakke) [inline src/hott/02-homotopies.rzk.md:85]:
```suggestion
 : dependent-homotopy
    ( A)
    ( \ b → g b = g b)
    ( \ x → concat B (g x) (f x) (g x) (rev B (f x) (g x) (K x)) (H x))
    ( refl-htpy A B g)
```

**Review by fredrik-bakke — COMMENTED** [review]
Sorry, here are some corrections on formatting as well.

By the way, are you planning on doing the refactoring of homotopies in this PR, or some other work? If so, it would be helpful if you marked it as a draft so it's easier for us to spot that it is unfinished.

File: src/hott/02-homotopies.rzk.md (hunk 6 above)
Comment (jonalfcam) [reply src/hott/02-homotopies.rzk.md:85]: Thanks for the formatting suggestions. And indeed there is quite a bit left to work on this PR. I’ve marked it as a draft.

**Comment by jonalfcam** [comment]
I have finished with the homotopy coherence I wanted to work on.  @fredrik-bakke if it is ok, I will update the `homotopy` and `dependent-homotopy`  in a PR of its own.  Also, thank you for recommending the rainbow-indent plugin!

**Comment by fredrik-bakke** [comment]
> I have finished with the homotopy coherence I wanted to work on. @fredrik-bakke if it is ok, I will update the `homotopy` and `dependent-homotopy` in a PR of its own. Also, thank you for recommending the rainbow-indent plugin!

That's fine! Maybe just rename `dependent-homotopy` to `dhomotopy` for now, so as not to keep the misleading name for it. And I'm glad the plugin was useful to you :)

**Comment by jonalfcam** [comment]
> That's fine! Maybe just rename `dependent-homotopy` to `dhomotopy` for now, so as not to keep the misleading name for it. And I'm glad the plugin was useful to you :)

Great! I'm made the change.

**Comment by emilyriehl** [comment]
I'm a bit slow to join this. It looks like this is ready to merge. Any outstanding issues I'm not aware of?

File: src/hott/01-paths.rzk.md [hunk 7]
```diff
@@ -380,6 +380,18 @@ The following needs to be outside the previous section because of the usage of
...
+A special case of this is sometimes useful
+
+```rzk
+#def cancel-left-path
+  (A : U)
+  (x y : A)
+  (p q : x = y)
+  : (p = q) → ( concat A y x y (rev A x y q) p) = refl
+  := triangle-rotation A x y y p q refl
+
```
Comment (fredrik-bakke) [inline src/hott/01-paths.rzk.md:392]:
```suggestion
```

File: src/hott/01-paths.rzk.md [hunk 8]
```diff
@@ -952,3 +964,329 @@ The following is the same as above but with alternating arguments.
...
+            ( concat-eq-right
+              ( A )
+              ( v )
+              ( w)
+              ( z)
+              ( p)
+              ( concat A w v z q r)
+              ( concat A w y z s t)
+              ( H)))))
+      ( rev-associative-concat A v w y z p s t))
```
Comment (fredrik-bakke) [inline src/hott/01-paths.rzk.md:1180]: Could this proof be simplified by applying the previous lemma and concatenation by the associator for path concatenation?

File: src/hott/01-paths.rzk.md [hunk 9]
```diff
@@ -952,3 +964,329 @@ The following is the same as above but with alternating arguments.
...
+  ( t : y = z)
+  ( H : (concat A w v z q r) = (concat A w y z s t))
+  ( H' : (concat A v w v p q) = refl)
+  : ( concat A v y z (concat A v w y p s) t) = r
+  :=
+    rev
+      ( v = z)
+      ( r)
+      ( concat A v y z (concat A v w y p s) t)
+      ( eq-top-cancel-commutative-square' A v w y z p q s r t H H')
```
Comment (fredrik-bakke) [inline src/hott/01-paths.rzk.md:1202]: We're trying out a new indentation scheme for the top level part of a definition, to save on horizontal space:
```suggestion
  :=
  rev
  ( v = z)
  ( r)
  ( concat A v y z (concat A v w y p s) t)
  ( eq-top-cancel-commutative-square' A v w y z p q s r t H H')
```

File: src/hott/02-homotopies.rzk.md [hunk 10]
```diff
@@ -113,6 +155,24 @@ unlike composition.
...
+  ( H : homotopy A B f g)
+  ( x y : A)
+  ( p : x = y)
+  : ( concat B (f x) (g x) (g y) (H x) (ap A B x y g p)) =
+    ( concat B (f x) (f y) (g y) (ap A B x y f p) (H y))
+  := rev
+      ( f x = g y)
+      ( concat B (f x) (f y) (g y) (ap A B x y f p) (H y))
+      ( concat B (f x) (g x) (g y) (H x) (ap A B x y g p))
+      ( nat-htpy A B f g H x y p)
```
Comment (fredrik-bakke) [inline src/hott/02-homotopies.rzk.md:173]: Always break line after walrus separator when definition is multiline
```suggestion
  :=
  rev
  ( f x = g y)
  ( concat B (f x) (f y) (g y) (ap A B x y f p) (H y))
  ( concat B (f x) (g x) (g y) (H x) (ap A B x y g p))
  ( nat-htpy A B f g H x y p)
```

File: src/hott/04-half-adjoint-equivalences.rzk.md [hunk 11]
```diff
@@ -804,3 +804,181 @@ have equivalent identity types by showing that equivalences are embeddings.
...
+      ( (map-inverse-is-half-adjoint-equiv A B f is-hae) (f a))
+      ( f)
+      ( rev ( A)
+        ( (map-inverse-is-half-adjoint-equiv A B f is-hae) (f a))
+        ( a)
+        ( (retraction-htpy-is-half-adjoint-equiv A B f is-hae) a)))
+    ((section-htpy-is-half-adjoint-equiv A B f is-hae) (f a))
+    (ap ( B) ( B) ( f a) ( b)
+      (\z → (f ((map-inverse-is-half-adjoint-equiv A B f is-hae) z)))
+      ( q))-- s
```
Comment (fredrik-bakke) [inline src/hott/04-half-adjoint-equivalences.rzk.md:972]:
```suggestion
      ( q))
```

**Review by fredrik-bakke — COMMENTED** [review]
There is still some formatting that could be improved in this contribution, but I will no longer comment on those. No offense meant to anyone, this is purely to preserve my own sanity and time.

Also, it should be checked that the diagrams render properly on the website. The correct approach would be to build the website locally and check it on your computer before the PR is merged. But I won't tell on anyone if the PR is merged and the diagrams are checked on the live webpage.

**Comment by fredrik-bakke** [comment]
Please let us know when you have addressed the comments, Jonathan

**Comment by emilyriehl** [comment]
@jonalfcam I see one unresolved conversation (where @fredrik-bakke suggested a potential simplification in the proof, which in any case is over-indented).

I have one other suggestion: throughout the sHoTT library it is more common to write markdown comments using code blocks than LaTeX (with the exception being cases where the code blocks are excessively ugly). Would you mind making this change where practical?

File: src/hott/01-paths.rzk.md (hunk 8 above)
Comment (jonalfcam) [reply src/hott/01-paths.rzk.md:1179]: Of course. And I have idea why I didn't do that. Thanks for pointing it out!

**Comment by jonalfcam** [comment]
> @jonalfcam I see one unresolved conversation (where @fredrik-bakke suggested a potential simplification in the proof, which in any case is over-indented).
> 
> I have one other suggestion: throughout the sHoTT library it is more common to write markdown comments using code blocks than LaTeX (with the exception being cases where the code blocks are excessively ugly). Would you mind making this change where practical?

Ah, my two great fondnesses: over-indentation and LaTeX. Just so I understand: the preference is for writing code rather than math in explanations?  Are commutative diagrams still ok?

**Comment by jonalfcam** [comment]
I might advocate for leaving some of the LaTeX in the coherences section. The idea behind including it was to explain some of the construction, and it seems appropriate for that. If we prefer not to have any LaTeX in the documentation, maybe we can revisit the zulip discussion about this.

**Comment by jonalfcam** [comment]
I think otherwise, everything is resolved now!

**Comment by emilyriehl** [comment]
> I might advocate for leaving some of the LaTeX in the coherences section. The idea behind including it was to explain some of the construction, and it seems appropriate for that. If we prefer not to have any LaTeX in the documentation, maybe we can revisit the zulip discussion about this.

Great idea to discuss this there. I'll merge now.

**Review by emilyriehl — APPROVED** [review]

**Pull request merged**